util/econftool.c

```diff
@@ -0,0 +1,557 @@
+    econf_error = econf_getGroups(key_file, &groupCount, &groups);
+    if (econf_error) {
+        fprintf(stderr, "%s\n", econf_errString(econf_error));
+        econf_free(groups);
+        econf_free(key_file);
+        return -1;
+    }
+    for (size_t g = 0; g < groupCount; g++)
+    {
+        char **keys = NULL;
+        size_t key_count = 0;
+
+        econf_error = econf_getKeys(key_file, groups[g], &key_count, &keys);
+        if (econf_error)
+        {
+            fprintf(stderr, "%s\n", econf_errString(econf_error));
+            econf_free(groups);
+            econf_free(keys);
+            return -1;
+        }
+        printf("%s\n", groups[g]);
+        for (size_t k = 0; k < key_count; k++) {
+            econf_error = econf_getStringValue(key_file, groups[g], keys[k], &value);
+            if (econf_error) {
+                fprintf(stderr, "%s\n", econf_errString(econf_error));
+                econf_free(groups);
+                econf_free(keys);
+                free(value);
+                return -1;
+            }
+            if (value != NULL)
+                printf("%s = %s\n", keys[k], value);
+            free(value);
+        }
+        printf("\n");
+        econf_free(keys);
+    }
+    econf_free(groups);
+    return 0;
+}
+
+/**
+ * @brief Generates a tmpfiles from key_file and opens editor to allow user editing.
+ *        It then saves the edited in key_file_edit and deletes the tmpfile
+ */
+static int econf_edit_editor(struct econf_file **key_file_edit)
+{
+    econf_err econf_error;
+    int ret = 0;
+    int error;
+    int wstatus;
+    char *editor;
+    char tmpfile_edit[FILENAME_MAX];
+    char path_tmpfile_edit[PATH_MAX];
+    char tmp_name[PATH_MAX];
+    char *tmp_path = getenv("TMPDIR");
+    if (tmp_path == NULL)
+        tmp_path = "/tmp";
+
+    if (generate_tmp_file(tmp_name))
+        return -1;
+    snprintf(path_tmpfile_edit, sizeof(path_tmpfile_edit), "%s", tmp_name);
+    snprintf(tmpfile_edit, sizeof(tmpfile_edit), "%s", strrchr(path_tmpfile_edit, '/') + 1);
+
+    editor = getenv("EDITOR");
+    if (editor == NULL) {
+        /* if no editor is specified take vi as default */
+        editor = "vi";
+    }
+
+    pid_t pid = fork();
+
+    if (pid == -1) {
+        perror("fork() failed");
+        exit(EXIT_FAILURE);
+    } else if (!pid) {
+    /* child */
+
+        /* write contents of key_file to tmpfile */
+        econf_error = econf_writeFile(key_file, tmp_path, tmpfile_edit);
+        if (econf_error) {
+            fprintf(stderr, "%s\n", econf_errString(econf_error));
+            exit(EXIT_FAILURE);
+        }
+        /* execute given command and save result in tmpfile_edit */
+        error = execlp(editor, editor, path_tmpfile_edit, (char *) NULL);
+
+        if (error != 0) {
+            fprintf(stderr, "Couldn't open %s! %s\n", editor, strerror(errno));
+            exit(EXIT_FAILURE);
+        } else {
+            exit(EXIT_SUCCESS);
```

Comment:
This can't actually happen - `execlp` only returns if an error occurred.

```
execlp(editor, editor, path_tmpfile_edit, (char *) NULL);
fprintf(stderr, "Couldn't open %s! %s\n", editor, strerror(errno));
exit(EXIT_FAILURE);
```